torch/csrc/autograd/functions/accumulate_grad.cpp

```diff
@@ -40,7 +40,20 @@ auto AccumulateGrad::apply(variable_list&& grads) -> variable_list {
 
   at::Tensor& grad = variable.grad();
   if (!grad.defined()) {
-    variable.grad() = new_grad.clone();
+    // under following condition, we can avoid clone()
+    if (!GradMode::is_enabled()
+        && !new_grad.type().is_sparse()
+        && new_grad.is_contiguous()
+        && new_grad.use_count() == 2) {
+      // first check it is in first-order grad only mode
+      // then check not sparse before is_contiguous
+      // then check contiguous, otherwise later in place accumulation may fail
+      // and lastly, check it is the last reference before we grab it
+      // we created new_grad above so ref count should be 2
```

Comment:
Can we move the variable into `new_grad`, or make it reference, such that we can assert that the use count is exactly 1?

Reply:
It seems to me we can make new_grad reference and assert use_count 1. It is actually better since this way we will also correctly catch new_grad that produced by hook with use_count==1.
Trying to modify grad in-place with hook will modify all grads sharing the buffer, what's your opinion on that? This behavior should not be changed by make new_grad reference though. So it is probably an unrelated question.

Reply:
Hooks should generally never modify their arguments, but we're not checking for that (doing so is UB)